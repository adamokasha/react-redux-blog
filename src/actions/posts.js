import axios from "axios";

export const setPosts = posts => ({
  type: "SET_POSTS",
  posts
});

export const startSetPosts = () => {
  return async dispatch => {
    try {
      const response = await axios.get(
        `${process.env.REACT_APP_API_URL}/posts`
      );
      dispatch(setPosts(response.data.posts));
    } catch (e) {
      throw new Error("Could not retrieve posts...");
    }
  };
};

export const addPost = post => ({
  type: "ADD_POST",
  post
});

export const startAddPost = post => {
  return async dispatch => {
    const { token } = JSON.parse(localStorage.getItem("auth"));
    try {
      const response = await axios({
        url: `${process.env.REACT_APP_API_URL}/posts`,
        method: "post",
        headers: {
          "x-auth": token
        },
        data: post
      });

      dispatch(addPost(response.data.post));
    } catch (e) {
      throw new Error(e.response.data.error);
    }
  };
};

export const editPost = updates => ({
  type: "EDIT_POST",
  updates
});

export const startEditPost = (post, id) => {
  return async dispatch => {
    const { token } = JSON.parse(localStorage.getItem("auth"));
    try {
      const response = await axios({
        url: `${process.env.REACT_APP_API_URL}/posts/${id}`,
        method: "patch",
        headers: {
          "x-auth": token
        },
        data: post
      });

      dispatch(editPost(response.data.post));
    } catch (e) {
      throw new Error(e.response.data.error);
    }
  };
};

export const deletePost = id => ({
  type: "DELETE_POST",
  id
});

export const startDeletePost = id => {
  return async dispatch => {
    const { token } = JSON.parse(localStorage.getItem("auth"));
    try {
      const response = await axios({
        url: `${process.env.REACT_APP_API_URL}/posts/${id}/`,
        method: "delete",
        headers: {
          "x-auth": token
        }
      });

      dispatch(deletePost(id));
    } catch (e) {
      throw new Error(e.response.data.error);
    }
  };
};

export const addComment = (id, comment) => ({
  type: "ADD_COMMENT",
  id,
  comment
});

export const startAddComment = (id, comment) => {
  return async dispatch => {
    try {
      const { token } = JSON.parse(localStorage.getItem("auth"));
      const response = await axios({
        url: `${process.env.REACT_APP_API_URL}/posts/${id}/comments`,
        method: "post",
        headers: {
          "x-auth": token
        },
        data: {
          comment
        }
      });

      dispatch(addComment(id, response.data));
    } catch (e) {
      throw new Error(e.response.data.error);
    }
  };
};
