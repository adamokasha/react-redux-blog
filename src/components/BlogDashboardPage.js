import React from "react";
import { connect } from "react-redux";

import Header from "./Header";
import PostsList from "./PostsList";

export const BlogDashboardPage = ({ displayName }) => {
  function renderWelcome() {
    if (displayName) {
      return <h3>Welcome {displayName}.</h3>;
    }
    return;
  }
  return (
    <div>
      <Header />
      <div id="landing">
        <div className="primary-overlay">
          <div className="container h-100">
            <div className="row hero-text">
              <div className="col-lg-7">
                <h1 className="text-white mb-5">
                  React Blog - Responsive blog styled with Bootstrap 4
                </h1>
                <a
                  href="https://github.com/adamokasha/react-redux-blog"
                  target="_blank"
                  className="text-uppercase text-white hero-link"
                >
                  <u>Learn More</u>
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <br />
      <div className="container mb-3">{renderWelcome()}</div>
      <div>
        <PostsList />
      </div>
    </div>
  );
};

const mapStateToProps = state => ({
  displayName: state.auth.displayName
});

export default connect(mapStateToProps)(BlogDashboardPage);
