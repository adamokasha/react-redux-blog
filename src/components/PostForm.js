import React from 'react';
import ErrorsList from './ErrorsList';

export class PostForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      _id: props.post ? props.post._id : '',
      title: props.post ? props.post.title : '',
      createdAt: props.post ? props.post.createdAt : Date.now(),
      category: props.post ? props.post.category : '',
      mainImage: props.post ? props.post.mainImage : '',
      thumbnail: props.post ? props.post.thumbnail : '',
      body: props.post ? props.post.body : '',
      errors: ''
    };
  }
  onTitleChange = e => {
    this.setState({ title: e.target.value });
  };
  onCategoryChange = e => {
    this.setState({ category: e.target.value });
  };
  onBodyChange = e => {
    this.setState({ body: e.target.value });
  };
  onMainImageChange = e => {
    this.setState({ mainImage: e.target.value });
  };
  onThumbnailChange = e => {
    this.setState({ thumbnail: e.target.value });
  };
  renderErrors = () => {
    const { errors } = this.state;
    if (errors) {
      return <ErrorsList errors={errors} />;
    }
  };
  onSubmit = e => {
    e.preventDefault();

    const { title, category, body } = this.state;
    const errors = [];
    if (title.length < 4) {
      errors.push('Title should have at least 4 characters.');
    }
    if (category.length < 4) {
      errors.push('Category should have at least 4 characters');
    }
    if (body.length < 24) {
      errors.push('The post body should have at least 24 characters');
    }

    if (errors.length >= 1) {
      this.setState({ errors });
    } else {
      const post = this.state;
      this.props.onSubmit(post);
    }
  };
  render() {
    return (
      <div>
        {this.renderErrors()}
        <form onSubmit={this.onSubmit}>
          <div className="form-group">
            <label htmlFor="title">Title</label>
            <input
              type="text"
              className="form-control"
              name="title"
              onChange={this.onTitleChange}
              value={this.state.title}
              maxLength={60}
            />
          </div>
          <div className="form-group">
            <label htmlFor="category">Category</label>
            <input
              type="text"
              className="form-control"
              name="category"
              onChange={this.onCategoryChange}
              value={this.state.category}
              maxLength={16}
            />
          </div>
          <div className="form-group">
            <label htmlFor="mainimage">Main Image (Optional)</label>
            <input
              type="text"
              className="form-control"
              name="mainimage"
              onChange={this.onMainImageChange}
              value={this.state.mainImage}
              placeholder="Enter URL for the main post view image"
            />
          </div>
          <div className="form-group">
            <label htmlFor="thumbnail">Image Thumbnail (Optional)</label>
            <input
              type="text"
              className="form-control"
              name="thumbnail"
              onChange={this.onThumbnailChange}
              value={this.state.thumbnail}
              placeholder="Enter URL for the dashboard card image"
            />
          </div>
          <div className="form-group">
            <label htmlFor="body">Body</label>
            <textarea
              name="body"
              className="form-control"
              onChange={this.onBodyChange}
              value={this.state.body}
              maxLength={12000}
              rows="10"
            />
          </div>
          <button className="btn btn-primary btn-block">Save Post</button>
        </form>
      </div>
    );
  }
}

export default PostForm;
