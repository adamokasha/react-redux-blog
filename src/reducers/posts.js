const postsReducer = (state = [], action) => {
  switch (action.type) {
    case "SET_POSTS":
      return action.posts;
    case "ADD_POST":
      return [...state, action.post];
    case "EDIT_POST":
      return state.map(post => {
        if (post._id === action.updates._id) {
          return { ...post, ...action.updates };
        } else {
          return post;
        }
      });
    case "ADD_COMMENT":
      return state.map(post => {
        if (post._id === action.id) {
          return {
            ...post,
            comments: [...post.comments, action.comment]
          };
        } else {
          return post;
        }
      });
    case "DELETE_POST":
      return state.filter(post => {
        if (post._id !== action.id) {
          return post;
        }
      });
    default:
      return state;
  }
};

export default postsReducer;
