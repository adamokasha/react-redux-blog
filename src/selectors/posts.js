export const sortPostsByDate = postsArray => {
  const newPostsArray = [].concat(postsArray);
  return newPostsArray.sort((a, b) => {
    if (a.createdAt > b.createdAt) {
      return -1;
    }
    if (a.createdAt < b.createdAt) {
      return 1;
    }
  });
};

export const sortCommentsByDate = commentsArray => {
  const newCommentsArray = [].concat(commentsArray);
  return newCommentsArray.sort((a, b) => {
    if (a.createdAt > b.createdAt) {
      return -1;
    }
    if (a.createdAt < b.createdAt) {
      return 1;
    }
  });
};
