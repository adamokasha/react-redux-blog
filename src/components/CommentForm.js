import React from "react";
import { connect } from "react-redux";

import { startAddComment } from "../actions/posts";

export class CommentForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      comment: "",
      error: ""
    };
  }
  onCommentChange = e => {
    this.setState({ comment: e.target.value });
  };
  renderError = () => {
    if (this.state.error) {
      return <div className="error-msg text-danger">{this.state.error}</div>;
    }
  };
  onSubmit = e => {
    e.preventDefault();

    const { comment } = this.state;
    if (comment.length < 8 || comment.length > 128) {
      this.setState({ error: "Comment must be between 8 and 128 characters." });
    } else {
      this.setState({ error: "" });
      this.props.startAddComment(this.props.id, this.state.comment);
      this.setState({ comment: "" });
    }
  };
  render() {
    return (
      <div className="mb-3">
        {this.renderError()}
        <form onSubmit={this.onSubmit}>
          <div className="form-group col-md-4 pl-0 mb-2">
            <textarea
              className="form-control"
              id="comment-textarea"
              placeholder="Leave a comment..."
              onChange={this.onCommentChange}
              maxLength={128}
              rows="3"
              value={this.state.comment}
            />
          </div>
          <button className="btn btn-primary">Add Comment</button>
        </form>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddComment: (id, comment) => dispatch(startAddComment(id, comment))
});

export default connect(
  null,
  mapDispatchToProps
)(CommentForm);
