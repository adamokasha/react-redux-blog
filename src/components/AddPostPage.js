import React from "react";
import { connect } from "react-redux";

import PostForm from "./PostForm";
import { startAddPost } from "../actions/posts";

export class AddPostPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      error: ""
    };
  }
  componentDidMount() {
    window.scrollTo(0, 0);
  }
  renderError = () => {
    if (this.state.error) {
      return (
        <div className="error-msg alert alert-danger" role="alert">
          <p>{this.state.error}</p>
        </div>
      );
    }
  };
  onSubmit = async post => {
    try {
      await this.props.startAddPost(post);
      this.props.history.push("/");
    } catch (e) {
      this.setState({ error: e.message });
    }
  };
  render() {
    return (
      <div className="container">
        <h3 className="text-center mt-4">Add Post</h3>
        {this.renderError()}
        <PostForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddPost: post => dispatch(startAddPost(post))
});

export default connect(
  null,
  mapDispatchToProps
)(AddPostPage);
