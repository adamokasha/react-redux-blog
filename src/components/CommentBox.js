import React from "react";
import { connect } from "react-redux";
import moment from "moment";

import CommentForm from "./CommentForm";
import { sortCommentsByDate } from "../selectors/posts";

export class CommentBox extends React.Component {
  renderCommentForm = () => {
    if (this.props.auth.id) {
      return <CommentForm id={this.props.id} />;
    }
    return <p className="font-italic">You must be signed in to comment...</p>;
  };
  renderComments = () => {
    if (this.props.comments.length > 0) {
      return sortCommentsByDate(this.props.comments).map(comment => {
        return (
          <li className="list-group-item pl-0" key={comment._id}>
            <p className="text-muted mb-1">
              Posted by {comment.createdBy} on{" "}
              {moment(comment.date).format("MM-DD-YY [at] HH:mm")}
            </p>
            <p className="mb-1">{comment.comment}</p>
          </li>
        );
      });
    } else {
      return <p>Be the first to comment...</p>;
    }
  };
  render() {
    return (
      <div>
        <h4>Comments</h4>
        {this.renderCommentForm()}
        <ul className="list-group-flush pl-0">{this.renderComments()}</ul>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  auth: state.auth
});

export default connect(mapStateToProps)(CommentBox);
