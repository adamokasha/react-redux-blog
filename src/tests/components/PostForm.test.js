import React from "react";
import { shallow, configure } from "enzyme";
import { mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import { PostForm } from "../../components/PostForm";
import posts from "../fixtures/posts";

configure({ adapter: new Adapter() });

test("should render PostForm correctly", () => {
  const wrapper = shallow(<PostForm />);
  expect(wrapper).toMatchSnapshot();
});

test("should render PostForm with post data", () => {
  const wrapper = shallow(<PostForm post={posts[0]} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render error if form input invalid", () => {
  const wrapper = mount(<PostForm />);
  wrapper.find("form").simulate("submit", {
    preventDefault: () => {}
  });
  expect(wrapper.state("errors")).toBeTruthy();
  expect(wrapper.find(".list-group-item-danger").length).toBeGreaterThan(0);
});

test("should set title on input change", () => {
  const value = "A Title";
  const wrapper = shallow(<PostForm />);
  wrapper
    .find("input")
    .at(0)
    .simulate("change", {
      target: { value }
    });
  expect(wrapper.state("title")).toBe(value);
});

test("should set category on input change", () => {
  const value = "A Category";
  const wrapper = shallow(<PostForm />);
  wrapper
    .find("input")
    .at(1)
    .simulate("change", {
      target: { value }
    });
  expect(wrapper.state("category")).toBe(value);
});

test("should set body on input change", () => {
  const value = "The body of the post of at least 24 characters";
  const wrapper = shallow(<PostForm />);
  wrapper.find("textarea").simulate("change", {
    target: { value }
  });
  expect(wrapper.state("body")).toBe(value);
});

test("should call onSubmit prop for valid form submission", () => {
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(<PostForm post={posts[1]} onSubmit={onSubmitSpy} />);
  wrapper.find("form").simulate("submit", {
    preventDefault: () => {}
  });
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    _id: posts[1]._id,
    title: posts[1].title,
    category: posts[1].category,
    mainImage: "",
    thumbnail: "",
    body: posts[1].body,
    createdAt: posts[1].createdAt,
    errors: ""
  });
});
