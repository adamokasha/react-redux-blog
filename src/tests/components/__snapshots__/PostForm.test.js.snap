// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PostForm correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="title"
      >
        Title
      </label>
      <input
        className="form-control"
        maxLength={60}
        name="title"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="category"
      >
        Category
      </label>
      <input
        className="form-control"
        maxLength={16}
        name="category"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="mainimage"
      >
        Main Image (Optional)
      </label>
      <input
        className="form-control"
        name="mainimage"
        onChange={[Function]}
        placeholder="Enter URL for the main post view image"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="thumbnail"
      >
        Image Thumbnail (Optional)
      </label>
      <input
        className="form-control"
        name="thumbnail"
        onChange={[Function]}
        placeholder="Enter URL for the dashboard card image"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="body"
      >
        Body
      </label>
      <textarea
        className="form-control"
        maxLength={12000}
        name="body"
        onChange={[Function]}
        rows="10"
        value=""
      />
    </div>
    <button
      className="btn btn-primary btn-block"
    >
      Save Post
    </button>
  </form>
</div>
`;

exports[`should render PostForm with post data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="title"
      >
        Title
      </label>
      <input
        className="form-control"
        maxLength={60}
        name="title"
        onChange={[Function]}
        type="text"
        value="The first post"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="category"
      >
        Category
      </label>
      <input
        className="form-control"
        maxLength={16}
        name="category"
        onChange={[Function]}
        type="text"
        value="General"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="mainimage"
      >
        Main Image (Optional)
      </label>
      <input
        className="form-control"
        name="mainimage"
        onChange={[Function]}
        placeholder="Enter URL for the main post view image"
        type="text"
        value="http://mysite.com/img.jpg"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="thumbnail"
      >
        Image Thumbnail (Optional)
      </label>
      <input
        className="form-control"
        name="thumbnail"
        onChange={[Function]}
        placeholder="Enter URL for the dashboard card image"
        type="text"
        value="http://mysite.com/thumb.jpg"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="body"
      >
        Body
      </label>
      <textarea
        className="form-control"
        maxLength={12000}
        name="body"
        onChange={[Function]}
        rows="10"
        value="This is the first post. The body has at least 24 characters."
      />
    </div>
    <button
      className="btn btn-primary btn-block"
    >
      Save Post
    </button>
  </form>
</div>
`;
